File: vault-android/app/src/androidTest/java/net/koofr/vault/tests/MemorySecureStorage.kt
package net.koofr.vault.tests

import net.koofr.vault.SecureStorage

class MemorySecureStorage : SecureStorage {
    private var data: HashMap<String, String> = HashMap()

    fun getData(): HashMap<String, String> {
        return HashMap(data)
    }

    @Synchronized
    override fun getItem(key: String): String? {
        return data[key]
    }

    @Synchronized
    override fun setItem(key: String, value: String) {
        data[key] = value
    }

    @Synchronized
    override fun removeItem(key: String) {
        data.remove(key)
    }

    @Synchronized
    override fun clear() {
        data.clear()
    }
}

File: vault-mobile/src/memory_secure_storage.rs
use std::{
    collections::HashMap,
    sync::{Arc, RwLock, RwLockReadGuard},
};

use crate::{SecureStorage, SecureStorageError};

#[derive(Debug)]
pub struct MemorySecureStorage {
    data: Arc<RwLock<HashMap<String, String>>>,
}

impl MemorySecureStorage {
    pub fn new() -> Self {
        Self::new_with_data(HashMap::new())
    }

    pub fn new_with_data(data: HashMap<String, String>) -> Self {
        Self {
            data: Arc::new(RwLock::new(data)),
        }
    }

    pub fn get_data(&self) -> RwLockReadGuard<'_, HashMap<String, String>> {
        self.data.read().unwrap()
    }
}

impl SecureStorage for MemorySecureStorage {
    fn get_item(&self, key: String) -> Result<Option<String>, SecureStorageError> {
        Ok(self.data.read().unwrap().get(&key).cloned())
    }

    fn set_item(&self, key: String, value: String) -> Result<(), SecureStorageError> {
        self.data.write().unwrap().insert(key, value);

        Ok(())
    }

    fn remove_item(&self, key: String) -> Result<(), SecureStorageError> {
        self.data.write().unwrap().remove(&key);

        Ok(())
    }

    fn clear(&self) -> Result<(), SecureStorageError> {
        self.data.write().unwrap().clear();

        Ok(())
    }
}
